graders/python/python_autograder/run.sh

```diff
@@ -0,0 +1,74 @@
+#! /bin/bash
+
+##########################
+# INIT
+##########################
+
+# the directory where the job stuff is
+JOB_DIR='/grade/'
+# the other directories
+STUDENT_DIR=$JOB_DIR'student/'
+AG_DIR='/python_autograder/'
+TEST_DIR=$JOB_DIR'tests/'
+OUT_DIR=$JOB_DIR'results/'
+
+# where we will copy everything
+MERGE_DIR=$JOB_DIR'run/'
+# where we will put the actual student code- this depends on what the autograder expects, etc
+
+# now set up the stuff so that our run.sh can work
+mkdir $MERGE_DIR
+mkdir $OUT_DIR
+
+mv $STUDENT_DIR* $MERGE_DIR
+mv $AG_DIR* $MERGE_DIR
+mv $TEST_DIR* $MERGE_DIR
+
+# user does not need a copy of this script
+rm -f "$MERGE_DIR/run.sh"
+rm -rf "$JOB_DIR/shared"
+
+# we need this to include code as python modules
+echo "" > $MERGE_DIR/__init__.py
+
+# Do not allow ag user to modify, rename, or delete any existing files
+chmod -R 755 "$MERGE_DIR"
+chmod 1777 "$MERGE_DIR"
+
+cd $MERGE_DIR
+
+# Create directory without sticky bit for deletable files
+mkdir filenames
+chmod 777 filenames
+mv ans.py setup_code.py test.py filenames
+
+##########################
+# RUN
+##########################
+
+echo "[run] starting autograder"
+
+# write name of the secret file to a file
+SECRET_NAME=`uuidgen`
+echo -n "$SECRET_NAME" > filenames/output-fname.txt
+chmod +r filenames/output-fname.txt
+
+# we run the autograder as a limited user called ag
+su -c 'python3 pltest.py' ag
+
+rm -f results.json
+rm -f "$OUT_DIR/results.json"
```

Comment:
Not that it really matters, but OUT_DIR already has a trailing slash. I actually think that's a bit weird. I prefer to define directory names without trailing slashes, and then add the slash when appending.